src/Components/Filters.js

```diff
@@ -1,54 +1,154 @@
       <Text style={styles.buttonText}>{props.title}</Text>
     </TouchableOpacity>
   );
 };
 
-const Filters = props => {
-  const [level, setLevel] = useState();
+const HeaderButton = props => {
+  return (
+    <TouchableOpacity onPress={props.onPress} testID="validateButton">
+      <Text style={styles.headerButtonText}>✓</Text>
+    </TouchableOpacity>
+  );
+};
 
-  const onLevelChange = pressedLevel => {
-    const newLevel = pressedLevel === level ? undefined : pressedLevel;
-    setLevel(newLevel);
-  };
+class Filters extends React.Component {
+  constructor(props) {
+    super(props);
 
-  useEffect(() => {
-    const newData = level ? props.initialData.filter(drill => drill.level === level) : props.initialData;
-    props.onFiltered(newData);
-  }, [level]);
+    this.state = {
+      selectedLevel: undefined,
+      selectedGoal: undefined,
+      numberOfPlayers: undefined,
+      displayedDrills: this.props.route.params.initialData,
+    };
 
-  return (
-    <View style={styles.filters}>
-      <Text style={styles.filterTitle}>Level</Text>
-      <View style={styles.filter}>
-        <Button title="Beginner" onPress={() => onLevelChange('beginner')} active={level === 'beginner'} />
-        <Button title="Intermediate" onPress={() => onLevelChange('intermediate')} active={level === 'intermediate'} />
-        <Button title="Advanced" onPress={() => onLevelChange('advanced')} active={level === 'advanced'} />
+    this.onNumberOfPlayersChange = this.onInputChange.bind(this, 'numberOfPlayers');
+  }
+
+  onPressedChange(target, value) {
+    this.setState(prevState => {
+      const newValue = value === prevState[target] ? undefined : value;
+      return {
+        [target]: newValue,
+      };
+    }, this.applyFilters);
+  }
+
+  onInputChange(target, value) {
+    this.setState({ [target]: value }, this.applyFilters);
+  }
+
+  applyFilters() {
+    const { selectedLevel, selectedGoal, numberOfPlayers } = this.state;
+    let newData = this.props.route.params.initialData;
+
+    if (selectedLevel) newData = newData.filter(drill => drill.level === selectedLevel);
+    if (selectedGoal) newData = newData.filter(drill => drill.goals.includes(selectedGoal));
+    if (numberOfPlayers) newData = newData.filter(drill => drill.minimalPlayersNumber <= numberOfPlayers);
+
+    this.setState({ displayedDrills: newData });
+  }
+
+  componentDidMount() {
+    this.props.navigation.setOptions({
+      headerRight: () => (
+        <HeaderButton
+          onPress={() => {
+            this.props.navigation.navigate(this.props.route.params.previousScreen, {
+              filteredDrills: this.state.displayedDrills,
+              type: this.props.route.params.previousType,
+            });
+          }}
+        />
+      ),
+    });
+  }
+
+  render() {
+    const { selectedLevel, selectedGoal, numberOfPlayers } = this.state;
+    return (
+      <View style={styles.wrapper}>
```

Comment:
To better readability and reuse, you could split this big `render` function in some shorter one. Well, this is up to you ;)

Reply:
I was thinking about refactoring the filters component. But I wanted to build the Fitness Filters before to get a wider idea of what could be the right abstraction (or even trainings filters to follow the rule of three)